File: ios/CourierExpoDelegate.m
//
//  CourierExpoDelegate.m
//  courier-react-native
//
//  Created by Michael Miller on 3/25/25.
//

#import "CourierExpoDelegate.h"

@import Courier_iOS;

#import <React/RCTBundleURLProvider.h>
#import <React/RCTLinkingManager.h>

@implementation CourierExpoDelegate

NSString *expoIosForegroundNotificationPresentationOptions = @"iosForegroundNotificationPresentationOptions";
NSUInteger expoNotificationPresentationOptions = UNNotificationPresentationOptionNone;

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
  // Set the Courier user agent
  Courier.agent = [CourierAgent reactNativeIOS:@"5.4.3"];
  
  // Register for remote notifications
  [application registerForRemoteNotifications];
  
  // Set the UNUserNotificationCenter delegate
  UNUserNotificationCenter *center = [UNUserNotificationCenter currentNotificationCenter];
  center.delegate = self;
  
  // Observe changes to the iOS foreground notification presentation options
  [[NSNotificationCenter defaultCenter]
      addObserver:self
      selector:@selector(notificationPresentationOptionsUpdate:)
      name:expoIosForegroundNotificationPresentationOptions
      object:nil
  ];
  
  // Standard Expo/RN setup
  self.moduleName = @"main";
  self.initialProps = @{};
  
  return [super application:application didFinishLaunchingWithOptions:launchOptions];
}

#pragma mark - Custom Courier / Notification Methods

- (void)notificationPresentationOptionsUpdate:(NSNotification *)notification
{
  if ([[notification name] isEqualToString:expoIosForegroundNotificationPresentationOptions]) {
    NSDictionary *userInfo = notification.userInfo;
    expoNotificationPresentationOptions = ((NSNumber *)[userInfo objectForKey:@"options"]).unsignedIntegerValue;
  }
}

- (void)userNotificationCenter:(UNUserNotificationCenter *)center willPresentNotification:(UNNotification *)notification withCompletionHandler:(void (^)(UNNotificationPresentationOptions))completionHandler
{
    
    UNNotificationContent *content = notification.request.content;
    NSDictionary *message = content.userInfo;
    
    [message trackMessageWithEvent:CourierTrackingEventDelivered completion:^(NSError * _Nullable error) {
        if (error) {
            NSLog(@"Error tracking message: %@", error.localizedDescription);
        }
    }];
    
    dispatch_async(dispatch_get_main_queue(), ^{
    
        NSDictionary *pushNotification = [Courier formatPushNotificationWithContent:content];
        [[NSNotificationCenter defaultCenter] postNotificationName:@"pushNotificationDelivered" object:nil userInfo:pushNotification];
        
        completionHandler(expoNotificationPresentationOptions);
        
    });
    
}

- (void)userNotificationCenter:(UNUserNotificationCenter *)center didReceiveNotificationResponse:(UNNotificationResponse *)response withCompletionHandler:(void (^)(void))completionHandler
{
    
    UNNotificationContent *content = response.notification.request.content;
    NSDictionary *message = content.userInfo;
    
    [message trackMessageWithEvent:CourierTrackingEventClicked completion:^(NSError * _Nullable error) {
        if (error) {
            NSLog(@"Error tracking message: %@", error.localizedDescription);
        }
    }];

    dispatch_async(dispatch_get_main_queue(), ^{

        NSDictionary *pushNotification = [Courier formatPushNotificationWithContent:content];
        [[NSNotificationCenter defaultCenter] postNotificationName:@"pushNotificationClicked" object:nil userInfo:pushNotification];

        completionHandler();

    });
    
}

- (void)application:(UIApplication *)application didFailToRegisterForRemoteNotificationsWithError:(NSError *)error
{
    NSLog(@"Failed to rgister for remote notification token: %@", error.localizedDescription);
}

- (void)application:(UIApplication *)application didRegisterForRemoteNotificationsWithDeviceToken:(NSData *)deviceToken
{
    [Courier setAPNSToken:deviceToken];
}

@end

File: example-expo-55/ios/CourierExpoExample/CourierExpoDelegate.swift
import UIKit
import UserNotifications
import Courier_iOS
import React

@objc class CourierExpoDelegate: NSObject, UNUserNotificationCenterDelegate {

  private var notificationPresentationOptions: UNNotificationPresentationOptions = []
  private var cachedMessage: [AnyHashable: Any]?
  private var isReactNativeReady = false

  private let courierForegroundOptionsDidChangeNotification = Notification.Name("iosForegroundNotificationPresentationOptions")

  override init() {
    super.init()

    Courier.agent = CourierAgent.reactNativeExpoIOS("5.8.0")

    UIApplication.shared.registerForRemoteNotifications()

    UNUserNotificationCenter.current().delegate = self

    NotificationCenter.default.addObserver(
      self,
      selector: #selector(notificationPresentationOptionsUpdate(_:)),
      name: courierForegroundOptionsDidChangeNotification,
      object: nil
    )

    NotificationCenter.default.addObserver(
      self,
      selector: #selector(onBridgeWillReload),
      name: NSNotification.Name.RCTBridgeWillReload,
      object: nil
    )

    NotificationCenter.default.addObserver(
      self,
      selector: #selector(onReactNativeReady(_:)),
      name: NSNotification.Name.RCTContentDidAppear,
      object: nil
    )
  }

  @objc private func onReactNativeReady(_ note: Notification) {
    isReactNativeReady = true

    if let message = cachedMessage {
      NotificationCenter.default.post(
        name: Notification.Name("pushNotificationClicked"),
        object: nil,
        userInfo: message
      )
      cachedMessage = nil
    }
  }

  @objc private func onBridgeWillReload() {
    isReactNativeReady = false
  }

  @objc private func notificationPresentationOptionsUpdate(_ notification: Notification) {
    if let userInfo = notification.userInfo,
       let options = userInfo["options"] as? NSNumber {
      notificationPresentationOptions = UNNotificationPresentationOptions(rawValue: options.uintValue)
    }
  }

  func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse, withCompletionHandler completionHandler: @escaping () -> Void) {
    let content = response.notification.request.content
    let message = content.userInfo

    Task {
      await message.trackMessage(event: .clicked)
    }

    let pushNotification = Courier.formatPushNotification(content: content)

    DispatchQueue.main.async { [weak self] in
      guard let self = self else { return }
      if self.isReactNativeReady {
        NotificationCenter.default.post(
          name: Notification.Name("pushNotificationClicked"),
          object: nil,
          userInfo: pushNotification
        )
      } else {
        self.cachedMessage = pushNotification
      }
      completionHandler()
    }
  }

  func userNotificationCenter(_ center: UNUserNotificationCenter, willPresent notification: UNNotification, withCompletionHandler completionHandler: @escaping (UNNotificationPresentationOptions) -> Void) {
    let content = notification.request.content
    let message = content.userInfo

    Task {
      await message.trackMessage(event: .delivered)
    }

    DispatchQueue.main.async { [weak self] in
      guard let self = self else { return }
      let pushNotification = Courier.formatPushNotification(content: content)
      NotificationCenter.default.post(
        name: Notification.Name("pushNotificationDelivered"),
        object: nil,
        userInfo: pushNotification
      )
      completionHandler(self.notificationPresentationOptions)
    }
  }

  @objc func application(_ application: UIApplication, didFailToRegisterForRemoteNotificationsWithError error: Error) {
    print("[Courier] Failed to register for remote notifications: \(error.localizedDescription)")
  }

  @objc func application(_ application: UIApplication, didRegisterForRemoteNotificationsWithDeviceToken deviceToken: Data) {
    Task { @CourierActor in
      Courier.setAPNSToken(deviceToken)
    }
  }

  deinit {
    cachedMessage = nil
    isReactNativeReady = false
    NotificationCenter.default.removeObserver(self)
  }
}
